cumulus/client/consensus/aura/src/collators/slot_based/block_builder_task.rs

```diff
@@ -387,3 +433,64 @@ where
 		}
 	}
 }
+
+/// Translate the slot of the relay parent to the slot of the parachain.
+fn adjust_para_to_relay_parent_slot(
+	relay_header: &RelayHeader,
+	relay_chain_slot_duration: Duration,
+	para_slot_duration: SlotDuration,
+) -> Option<SlotInfo> {
+	let relay_slot = sc_consensus_babe::find_pre_digest::<RelayBlock>(&relay_header)
+		.map(|babe_pre_digest| babe_pre_digest.slot())
+		.ok()?;
+	let new_slot = Slot::from_timestamp(
+		relay_slot
+			.timestamp(SlotDuration::from_millis(relay_chain_slot_duration.as_millis() as u64))
+			.unwrap(),
+		para_slot_duration,
+	);
+	let para_slot = SlotInfo { slot: new_slot, timestamp: new_slot.timestamp(para_slot_duration)? };
+	tracing::debug!(
+		target: LOG_TARGET,
+		timestamp = ?para_slot.timestamp,
+		slot = ?para_slot.slot,
+		"Parachain slot adjusted to relay chain.",
+	);
+	Some(para_slot)
+}
+
+/// Traverses the ancestry of the given relay chain header to find the relay parent at the correct
+/// offset.
+async fn find_relay_parent_with_offset<RelayClient>(
```

Comment:
I find it weird that you return the relay parent at the offset twice (once the header and then part of the Vec).

Also I find it weird that the list is ordered ascending and not descending.

Reply:
I experimented with the order a bit during implementation, but it does not make a big difference. I am starting to check from the RP because there I can use the normale authorities for verification. If I later find an epoch change, I will use the new ones going forward. If I iterate from the back would be slightly more annoying, but in the end its all the same :man_shrugging: